commit d93b0c37287b63a5ad5241838e1b9be07b79e78a
Author:     Yuhong Wen <yh_wen@hotmail.com>
AuthorDate: Wed Mar 30 10:57:31 2022 -0400
Commit:     GitHub <noreply@github.com>
CommitDate: Wed Mar 30 10:57:31 2022 -0400

    Runner process (#369)
    
    * Clean up codes in nvflare/private/fed/app
    
    * Fix issues
    
    * Undo removal
    
    * Move server deployer
    
    * Fix format issue
    
    * Test server sub_process.
    
    * Enabled the server side child process for run.
    
    * temp changes.
    
    * server child process for training.
    
    * Fixed run_number.
    
    * Set the server to stopped after training complete.
    
    * thread safe for server child process communication.
    
    * Move into deployer folder
    
    * Fix import path
    
    * Use relative import
    
    * Communicator merged.
    
    * Merged with StudyManagerSpec.
    
    * Removed theStudyManagerSpec.
    
    Co-authored-by: YuanTingHsieh <yuantinghsieh@gmail.com>
---
 nvflare/private/fed/server/server_engine.py | 154 ++++++++++++++++++----------
 1 file changed, 98 insertions(+), 56 deletions(-)

diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index b8385576..01eeda38 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -13,25 +13,29 @@
 # limitations under the License.
 
 import copy
-import gc
 import logging
 import os
 import pickle
 import re
+import shlex
 import shutil
+import subprocess
 import sys
+import threading
 import time
-import traceback
 from concurrent.futures import ThreadPoolExecutor
+from multiprocessing.connection import Client as CommandCLient
 from threading import Lock
 from typing import List, Tuple
 
 from nvflare.apis.client import Client
-from nvflare.apis.fl_constant import FLContextKey, MachineStatus, ReservedTopic, ReturnCode, SnapshotKey
+from nvflare.apis.fl_constant import FLContextKey, MachineStatus, ReservedTopic, ReturnCode, SnapshotKey, \
+    AdminCommandNames, ServerCommandNames, ServerCommandKey
 from nvflare.apis.fl_context import FLContext
 from nvflare.apis.fl_snapshot import FLSnapshot
 from nvflare.apis.impl.job_def_manager import SimpleJobDefManager
 from nvflare.apis.shareable import Shareable, make_reply
+from nvflare.apis.utils.common_utils import get_open_ports
 from nvflare.apis.study_manager_spec import StudyManagerSpec
 from nvflare.apis.utils.fl_context_utils import get_serializable_data
 from nvflare.apis.workspace import Workspace
@@ -41,7 +45,6 @@ from nvflare.private.admin_defs import Message
 from nvflare.private.fed.server.server_json_config import ServerJsonConfigurator
 from nvflare.widgets.info_collector import InfoCollector
 from nvflare.widgets.widget import Widget, WidgetID
-
 from .client_manager import ClientManager
 from .run_manager import RunManager
 from .server_engine_internal_spec import EngineInfo, RunInfo, ServerEngineInternalSpec
@@ -88,6 +91,8 @@ class ServerEngine(ServerEngineInternalSpec):
 
         self.asked_to_stop = False
         self.snapshot_persistor = snapshot_persistor
+        self.command_conn = None
+        self.child_process = None
 
     def _get_server_app_folder(self):
         return "app_server"
@@ -159,31 +164,64 @@ class ServerEngine(ServerEngineInternalSpec):
                 return "Server app does not exist. Please deploy the server app before starting."
 
             self.engine_info.status = MachineStatus.STARTING
+
+            app_custom_folder = ""
             if self.server.enable_byoc:
                 app_custom_folder = os.path.join(app_root, "custom")
-                try:
-                    sys.path.index(app_custom_folder)
-                except ValueError:
-                    self.remove_custom_path()
-                    sys.path.append(app_custom_folder)
-
-            # future = self.executor.submit(start_server_training, (self.server))
-            future = self.executor.submit(
-                lambda p: start_server_training(*p), [self.server, self.args, app_root, self.run_number, snapshot]
-            )
+            self.child_process = self._start_runner_process(self.args, app_root, self.run_number, app_custom_folder, snapshot)
 
-            start = time.time()
-            # Wait for the server App to start properly
-            while self.engine_info.status == MachineStatus.STARTING:
-                time.sleep(0.3)
-                if time.time() - start > 300.0:
-                    return "Could not start the server app."
+            threading.Thread(target=self.wait_for_complete).start()
 
-            if self.engine_info.status != MachineStatus.STARTED:
-                return f"Failed to start server app: {self.engine_info.status}"
+            if self.command_conn:
+                with self.lock:
+                    self.command_conn.close()
+                    self.command_conn = None
 
+            self.engine_info.status = MachineStatus.STARTED
             return ""
 
+    def wait_for_complete(self):
+        while True:
+            try:
+                self.get_command_conn()
+                if self.command_conn:
+                    with self.lock:
+                        data = {ServerCommandKey.COMMAND: ServerCommandNames.HEARTBEAT, ServerCommandKey.DATA: {}}
+                        self.command_conn.send(data)
+                time.sleep(1.0)
+            except BaseException:
+                self.child_process = None
+                self.command_conn = None
+                self.engine_info.status = MachineStatus.STOPPED
+                break
+
+    def _start_runner_process(self, args, app_root, run_number, app_custom_folder, snapshot):
+        new_env = os.environ.copy()
+        if app_custom_folder != "":
+            new_env["PYTHONPATH"] = new_env["PYTHONPATH"] + ":" + app_custom_folder
+
+        self.open_port = get_open_ports(1)[0]
+        if snapshot:
+            restore_snapshot = True
+        else:
+            restore_snapshot = False
+        command_options = ""
+        for t in args.set:
+            command_options += " " + t
+        command = (
+            f"{sys.executable} -m nvflare.private.fed.app.server.runner_process -m "
+            + args.workspace
+            + " -s fed_server.json -r " + app_root
+            + " -n " + str(run_number)
+            + " -p " + str(self.open_port)
+            + " --set" + command_options + " print_conf=True restore_snapshot=" + str(restore_snapshot)
+        )
+        # use os.setsid to create new process group ID
+
+        # command = f"{sys.executable} -m nvflare.private.fed.app.server.worker_process"
+        process = subprocess.Popen(shlex.split(command, " "), preexec_fn=os.setsid, env=new_env)
+        return process
+
     def remove_custom_path(self):
         regex = re.compile(".*/run_.*/custom")
         custom_paths = list(filter(regex.search, sys.path))
@@ -208,7 +246,22 @@ class ServerEngine(ServerEngineInternalSpec):
         self.logger.info("Abort the server app run.")
         # self.server.stop_training()
         self.server.shutdown = True
-        self.server.abort_run()
+        # self.server.abort_run()
+
+        try:
+            self.get_command_conn()
+            if self.command_conn:
+                with self.lock:
+                    data = {ServerCommandKey.COMMAND: AdminCommandNames.ABORT, ServerCommandKey.DATA: {}}
+                    self.command_conn.send(data)
+                    status_message = self.command_conn.recv()
+                    self.logger.info(f"Abort server: {status_message}")
+        except:
+            if self.child_process:
+                self.child_process.terminate()
+                self.child_process = None
+
+        self.engine_info.status = MachineStatus.STOPPED
         return ""
 
     def check_app_start_readiness(self) -> str:
@@ -294,6 +347,20 @@ class ServerEngine(ServerEngineInternalSpec):
         data = zip_directory_to_bytes(fullpath_src, "")
         return "", data
 
+    def get_app_run_info(self) -> RunInfo:
+        run_info = None
+        try:
+            self.get_command_conn()
+            if self.command_conn:
+                with self.lock:
+                    data = {ServerCommandKey.COMMAND: ServerCommandNames.GET_RUN_INFO, ServerCommandKey.DATA: {}}
+                    self.command_conn.send(data)
+                    run_info = self.command_conn.recv()
+        except:
+            pass
+
+        return run_info
+
     def set_run_manager(self, run_manager: RunManager):
         self.run_manager = run_manager
         for _, widget in self.widgets.items():
@@ -407,6 +474,14 @@ class ServerEngine(ServerEngineInternalSpec):
 
         return results
 
+    def get_command_conn(self):
+        if not self.command_conn:
+            try:
+                address = ("localhost", self.open_port)
+                self.command_conn = CommandCLient(address, authkey="client process secret password".encode())
+            except Exception as e:
+                pass
+
     def persist_components(self, fl_ctx: FLContext, completed: bool):
 
         # Call the State Persistor to persist all the component states
@@ -446,39 +521,6 @@ class ServerEngine(ServerEngineInternalSpec):
         self.executor.shutdown()
 
 
-def start_server_training(server, args, app_root, run_number, snapshot):
-
-    restart_file = os.path.join(args.workspace, "restart.fl")
-    if os.path.exists(restart_file):
-        os.remove(restart_file)
-
-    try:
-        server_config_file_name = os.path.join(app_root, args.server_config)
-
-        conf = ServerJsonConfigurator(
-            config_file_name=server_config_file_name,
-        )
-        conf.configure()
-
-        set_up_run_config(server, conf)
-
-        server.start_run(run_number, app_root, conf, args, snapshot)
-    except BaseException as e:
-        traceback.print_exc()
-        logging.getLogger().warning("FL server execution exception: " + str(e))
-    finally:
-        server.status = ServerStatus.STOPPED
-        server.engine.engine_info.status = MachineStatus.STOPPED
-        server.stop_training()
-        # if trainer:
-        #     trainer.close()
-
-        # Force garbage collection
-        gc.collect()
-
-    # return server.start()
-
-
 def server_shutdown(server, touch_file):
     with open(touch_file, "a"):
         os.utime(touch_file, None)
